File: fe/mobile/android/app/src/main/java/com/capstone/data/repository/AuthRepositoryImpl.java
package com.capstone.data.repository;

import com.capstone.data.datasource.AuthRemoteDataSource;
import com.capstone.data.source.request.ChangePasswordRequest;
import com.capstone.data.source.request.ResetPasswordRequest;
import com.capstone.data.source.request.UpdateProfileRequest;
import com.capstone.domain.model.UserProfile;
import com.capstone.domain.repository.AuthRepository;
import com.capstone.infrastructure.security.AntiBruteForceManager;
import com.capstone.infrastructure.security.TokenManager;

import androidx.annotation.Nullable;

public class AuthRepositoryImpl implements AuthRepository {
    private final AuthRemoteDataSource remote;
    private final TokenManager tokenManager;
    private final AntiBruteForceManager bruteForceManager;

    public AuthRepositoryImpl(AuthRemoteDataSource remote, TokenManager tokenManager, AntiBruteForceManager bruteForceManager) {
        this.remote = remote;
        this.tokenManager = tokenManager;
        this.bruteForceManager = bruteForceManager;
    }

    @Override
    public UserProfile login(String accessToken) throws Exception {
        var profileResponse = remote.login(accessToken);
        tokenManager.saveSession(accessToken, "SIMULATED_REFRESH_TOKEN", profileResponse.getRole());
        return UserProfile.fromResponse(profileResponse);
    }

    @Override
    public String sendOtp(String email) throws Exception {
        return remote.sendOtp(email);
    }

    @Override
    public String verifyOtp(String email, String otp) throws Exception {
        if (bruteForceManager.isLocked(email)) {
            throw new Exception("Tài khoản bị tạm khóa do thử sai quá nhiều lần. Vui lòng quay lại sau.");
        }
        try {
            var result = remote.verifyOtp(email, otp);
            bruteForceManager.resetAttempts(email);
            return result;
        } catch (Exception e) {
            bruteForceManager.recordFailure(email);
            throw e;
        }
    }

    @Override
    public String resetPassword(String email, String otp, String newPassword) throws Exception {
        if (bruteForceManager.isLocked(email)) {
            throw new Exception("Tài khoản đang bị khóa.");
        }
        try {
            var result = remote.resetPassword(new ResetPasswordRequest(email, otp, newPassword));
            bruteForceManager.resetAttempts(email);
            return result;
        } catch (Exception e) {
            bruteForceManager.recordFailure(email);
            throw e;
        }
    }

    @Override
    public String changePassword(String oldPass, String newPass, String confirmPass) throws Exception {
        return remote.changePassword(new ChangePasswordRequest(oldPass, newPass, confirmPass));
    }

    @Override
    public UserProfile updateProfile(@Nullable String fullName, @Nullable String username, @Nullable String phoneNumber,
                                     @Nullable String birthdate, @Nullable String address, @Nullable Boolean gender) throws Exception {
        var profileResponse = remote.updateProfile(
                new UpdateProfileRequest(fullName, username, phoneNumber, birthdate, address, gender)
        );
        return UserProfile.fromResponse(profileResponse);
    }

    @Override
    public UserProfile updateAvatar(byte[] imageBytes) throws Exception {
        var profileResponse = remote.updateAvatar(imageBytes);
        return UserProfile.fromResponse(profileResponse);
    }

    @Override
    public UserProfile getMe() throws Exception {
        var profileResponse = remote.getMe();
        return UserProfile.fromResponse(profileResponse);
    }
}

File: fe/mobile/android/app/src/main/java/com/capstone/data/repository/AuthRepositoryImpl.kt
package com.capstone.data.repository

import com.capstone.data.datasource.AuthRemoteDataSource
import com.capstone.domain.model.UserProfile
import com.capstone.domain.repository.AuthRepository
import com.capstone.infrastructure.security.TokenManager
import com.capstone.infrastructure.security.AntiBruteForceManager

class AuthRepositoryImpl(
    private val remote: AuthRemoteDataSource,
    private val tokenManager: TokenManager,
    private val bruteForceManager: AntiBruteForceManager
): AuthRepository {
    /**
     * Xử lý logic đăng nhập: Gọi DataSource để lấy Profile sau đó convert sang Domain Model.
     */
    override suspend fun login(accessToken: String): UserProfile {
        return remote.login(accessToken).let { profile ->
            tokenManager.saveSession(accessToken, "SIMULATED_REFRESH_TOKEN", profile.role)
            UserProfile(
                fullName = profile.fullName,
                avatarUrl = profile.avatarUrl,
                role = profile.role,
                username = profile.username,
                email = profile.email,
                departmentId = profile.departmentId,
                departmentName = profile.departmentName
            )
        }
    }

    /**
     * Gửi yêu cầu OTP thông qua Remote DataSource.
     */
    override suspend fun sendOtp(email: String): String {
        return remote.sendOtp(email)
    }

    /**
     * Xác thực mã OTP thông qua Remote DataSource.
     * Bảo mật: Thăng cường chống Brute Force.
     */
    override suspend fun verifyOtp(email: String, otp: String): String {
        if (bruteForceManager.isLocked(email)) {
            throw Exception("Tài khoản bị tạm khóa do thử sai quá nhiều lần. Vui lòng quay lại sau.")
        }
        return try {
            val result = remote.verifyOtp(email, otp)
            bruteForceManager.resetAttempts(email)
            result
        } catch (e: Exception) {
            bruteForceManager.recordFailure(email)
            throw e
        }
    }

    /**
     * Thực hiện đặt lại mật khẩu thông qua Remote DataSource.
     * Bảo mật: Thăng cường chống Brute Force.
     */
    override suspend fun resetPassword(email: String, otp: String, newPassword: String): String {
        if (bruteForceManager.isLocked(email)) {
            throw Exception("Tài khoản đang bị khóa.")
        }
        return try {
            val result = remote.resetPassword(com.capstone.data.source.request.ResetPasswordRequest(email, otp, newPassword))
            bruteForceManager.resetAttempts(email)
            result
        } catch (e: Exception) {
            bruteForceManager.recordFailure(email)
            throw e
        }
    }

    /**
     * Thực hiện đổi mật khẩu khi đã đăng nhập.
     */
    override suspend fun changePassword(oldPass: String, newPass: String, confirmPass: String): String {
        return remote.changePassword(com.capstone.data.source.request.ChangePasswordRequest(oldPass, newPass, confirmPass))
    }

    override suspend fun updateProfile(
        fullName: String?,
        username: String?,
        phoneNumber: String?,
        birthdate: String?,
        address: String?,
        gender: Boolean?
    ): UserProfile {
        return remote.updateProfile(com.capstone.data.source.request.UpdateProfileRequest(fullName, username, phoneNumber, birthdate, address, gender))
            .let { profile ->
                with(UserProfile) { profile.toDomain() }
            }
    }

    override suspend fun updateAvatar(imageBytes: ByteArray): UserProfile {
        return remote.updateAvatar(imageBytes).let { profile ->
            with(UserProfile) { profile.toDomain() }
        }
    }

    /**
     * Lấy thông tin tài khoản hiện tại và convert sang Domain Model.
     * Token được đính kèm tự động qua tầng Interceptor.
     */
    override suspend fun getMe(): UserProfile {
        return remote.getMe().let { profile ->
            with(UserProfile) { profile.toDomain() }
        }
    }
}